### PR TITLE
Introduce SslStreamBuilder

### PR DESCRIPTION
**Pull request by Ralith**
Draft a new API to support `SSL_stateless` and to allow for future early I/O support as previously discussed in #835.

### CODE REVIEW

File: openssl/src/ssl/mod.rs [hunk 1]
```diff
@@ -2796,6 +2766,109 @@ impl<S: Read + Write> Write for SslStream<S> {
     }
 }
 
+/// A partially constructed `SslStream`, useful for unusual handshakes.
+pub struct SslStreamBuilder<S> {
+    inner: Option<SslStream<S>>
```
Comment (sfackler) [inline openssl/src/ssl/mod.rs]: There's no need for this to be an option, right?

File: openssl/src/ssl/mod.rs [hunk 2]
```diff
@@ -2796,6 +2766,109 @@ impl<S: Read + Write> Write for SslStream<S> {
...
+    ///
+    /// This corresponds to [`SSL_stateless`]
+    ///
+    /// [`SSL_stateless`]: https://www.openssl.org/docs/manmaster/man3/SSL_stateless.html
+    #[cfg(ossl111)]
+    pub fn stateless(&mut self) -> Result<(), Error> {
+        if (unsafe { ffi::SSL_stateless(self.inner.as_mut().unwrap().ssl.as_ptr()) }) != 0 {
+            Ok(())
+        } else {
+            Err(self.inner.as_mut().unwrap().make_error(0))
```
Comment (sfackler) [inline openssl/src/ssl/mod.rs]: make_error needs to be called with the actual return value of SSL_stateless.

**Comment by sfackler** [comment]
Could you add a test using SSL_stateless?

File: openssl/src/ssl/mod.rs (hunk 2 above)
Comment (Ralith) [reply openssl/src/ssl/mod.rs]: That is the actual return value.

**Comment by Ralith** [comment]
Is there an in-memory stream type in the crate already somewhere that I can use for synchronous, deterministic testing? `SSL_stateless` isn't well-behaved under the "two threads/processes communicating over TCP" model that most of the other tests seem to use.

The function seems to be working well in the project I added it for, at least.

**Comment by Ralith** [comment]
Yep. In fact, that's exactly the structure I'm already using to bridge to OpenSSL in my project. I'll add something along those lines for the test.

File: openssl/src/ssl/mod.rs [hunk 3]
```diff
@@ -2796,6 +2766,109 @@ impl<S: Read + Write> Write for SslStream<S> {
...
+    ///
+    /// This corresponds to [`SSL_stateless`]
+    ///
+    /// [`SSL_stateless`]: https://www.openssl.org/docs/manmaster/man3/SSL_stateless.html
+    #[cfg(ossl111)]
+    pub fn stateless(&mut self) -> Result<(), Error> {
+        if (unsafe { ffi::SSL_stateless(self.inner.ssl.as_ptr()) }) != 0 {
+            Ok(())
+        } else {
+            Err(self.inner.make_error(0))
```
Comment (Ralith) [inline openssl/src/ssl/mod.rs]: Given that SSL_stateless doesn't seem to have interesting return values, should this go straight to `ErrorStack::get()` instead?

**Comment by Ralith** [comment]
After confirming that `SSL_get_error` does nothing relevant here, I went with `Option<ErrorStack>` for the corrected error type, as `SSL_stateless` empirically produces empty error stacks under the normal condition of sending a `HelloRetryReqest` in response to a `ClientHello` with no cookie. This enables callers to easily distinguish true error conditions (such as a malformed or inappropriate message) from routine operation.

**Comment by sfackler** [comment]
There are some contexts elsewhere where you can encounter errors with empty stacks - for example AEAD failures during symmetric decryption. My gut feeling is that we may just want to return a `Result<(), ErrorStack>` and document what an empty stack implies?

**Comment by sfackler** [comment]
Oh, but here the empty error stack case isn't really an error at all per se... hmmm

**Comment by Ralith** [comment]
I've opened an issue to discuss the ambiguity of this condition at https://github.com/openssl/openssl/issues/5549.

**Comment by mattcaswell** [comment]
Fix for this issue in openssl/openssl#5562

**Comment by sfackler** [comment]
Awesome, thanks!

**Comment by mattcaswell** [comment]
> Fix for this issue in openssl/openssl#5562

The above fix has been merged and is now available in OpenSSL master.

**Comment by Ralith** [comment]
Got a test written. Once 1.1.1-pre3 comes out (next week?) I'll update for the fixed return code, and swap in the new cookie callback functions for the old ones that have now been relegated to DTLSv1 use only.

**Comment by sfackler** [comment]
Sounds good!

**Comment by mattcaswell** [comment]
>  Once 1.1.1-pre3 comes out (next week?)

Due out 20th March.

**Comment by sfackler** [comment]
1.1.1-pre3 is out @Ralith

**Comment by Ralith** [comment]
Not sure what's going on with AppVeyor; looks spurious.

File: openssl-sys/src/ossl111.rs [hunk 4]
```diff
@@ -65,4 +65,20 @@ extern "C" {
                                   parse_arg: *mut c_void) -> c_int;
     pub fn SSL_stateless(s: *mut ::SSL) -> c_int;
     pub fn SSL_CIPHER_get_handshake_digest(cipher: *const ::SSL_CIPHER) -> *const ::EVP_MD;
+    pub fn SSL_CTX_set_stateless_cookie_generate_cb(
+        s: *mut ::SSL_CTX,
+        cb: Option<extern "C" fn(
```
Comment (sfackler) [inline openssl-sys/src/ossl111.rs]: This and below should be `unsafe extern "C" fn(...)`

File: openssl/src/ssl/mod.rs [hunk 5]
```diff
@@ -1437,8 +1437,9 @@ impl SslContextBuilder {
     /// The callback will be called with the SSL context and a slice into which the cookie
     /// should be written. The callback should return the number of bytes written.
     ///
-    /// This corresponds to `SSL_CTX_set_cookie_generate_cb`.
-    pub fn set_cookie_generate_cb<F>(&mut self, callback: F)
+    /// This corresponds to `SSL_CTX_set_stateless_cookie_generate_cb`.
+    #[cfg(ossl111)]
```
Comment (sfackler) [inline openssl/src/ssl/mod.rs]: Why is this no longer supported on older versions?

File: openssl/src/ssl/mod.rs [hunk 6]
```diff
@@ -2907,6 +2879,114 @@ impl<S: Read + Write> Write for SslStream<S> {
...
+    /// read, in which case the handshake should be continued via `accept`. If a
+    /// HelloRetryRequest containing a fresh cookie was transmitted, `Err(None)`
+    /// is returned instead. If the handshake cannot proceed at all,
+    /// `Err(Some(...))` is returned.
+    ///
+    /// This corresponds to [`SSL_stateless`]
+    ///
+    /// [`SSL_stateless`]: https://www.openssl.org/docs/manmaster/man3/SSL_stateless.html
+    #[cfg(ossl111)]
+    pub fn stateless(&mut self) -> Result<(), Option<ErrorStack>> {
```
Comment (sfackler) [inline openssl/src/ssl/mod.rs]: The convention elsewhere for this kind of thing is `Result<bool, ErrorStack>`.

File: openssl/src/ssl/mod.rs (hunk 5 above)
Comment (Ralith) [reply openssl/src/ssl/mod.rs]: `SSL_CTX_set_stateless_cookie_generate_cb` is new in OpenSSL 1.1.1-pre3. In -pre1, a DTLSv1-related interface was overloaded, but this was deemed suboptimal.

Comment (sfackler) [reply openssl/src/ssl/mod.rs]: So these methods are already part of the public API, so I don't think we can get rid of them. Seems fine to have both the old DTLSv1 and new TLSv1.3 stuff at the same time though? https://docs.rs/openssl/0.10.5/openssl/ssl/struct.SslContextBuilder.html#method.set_cookie_verify_cb

Comment (Ralith) [reply openssl/src/ssl/mod.rs]: Easy enough to support both here, but in general breaking changes between pre-release versions are expected.

Comment (sfackler) [reply openssl/src/ssl/mod.rs]: Someone using `set_cookie_verify_cb` and linking against OpenSSL 1.0.2 or 1.1.0 isn't using a prerelease version of anything AFAIKT.

Comment (Ralith) [reply openssl/src/ssl/mod.rs]: Fair enough. Anyway, it's all there now.

**Comment by sfackler** [comment]
Thanks!

**Pull request merged**